stripcaged.lua

```diff
@@ -0,0 +1,228 @@
+    for _, cage in ipairs(list) do
+        local count = {}
+        for _, ref in ipairs(cage.general_refs) do
+            if df.general_ref_contains_itemst:is_instance(ref) then
+                local classname = df.item_type.attrs[
+                    df.item.find(ref.item_id):getType()].caption
+                count[classname] = (count[classname] or 0) + 1
+            elseif df.general_ref_contains_unitst:is_instance(ref) then
+                local inventory = df.unit.find(ref.unit_id).inventory
+                for _, it in ipairs(inventory) do
+                    local classname = df.item_type.attrs[it.item:getType()].caption
+                    count[classname] = (count[classname] or 0) + 1
+                end
+                -- TODO: vermin ?
+
+                --[[ TODO: Determine how/if to handle a DEBUG flag.
+
+                --Ruby:
+                  else
+                      puts "unhandled ref #{ref.inspect}" if $DEBUG
+                  end
+                ]]
+            end
+        end
+
+        local type = df.item_type.attrs[cage:getType()].caption -- Default case
+        if df.item_cagest:is_instance(cage) then
+            type = 'Cage'
+        elseif df.item_animaltrapst:is_instance(cage) then
+            type = 'Animal trap'
+        end
+
+        -- If count is empty
+        if not next(count) then
+            empty_cages = empty_cages + 1
+        else
+            local sortedlist = {}
+            for classname, n in pairs(count) do
+                sortedlist[#sortedlist + 1] = {classname = classname, count = n}
+            end
+            table.sort(sortedlist, (function(i, j) return i.count < j.count end))
+            print(('%s %d: '):format(type, cage.id))
+            for _, t in ipairs(sortedlist) do
+                print(' ' .. t.count .. ' ' .. t.classname)
+            end
+        end
+        for k, v in pairs(count) do count_total[k] = (count_total[k] or 0) + v end
+    end
+
+    if #list > 1 then
+        print '\nTotal:'
+        local sortedlist = {}
+        for classname, n in pairs(count_total) do
+            sortedlist[#sortedlist + 1] = {classname = classname, count = n}
+        end
+        table.sort(sortedlist, (function(i, j) return i.count < j.count end))
+        for _, t in ipairs(sortedlist) do
+            print(' ' .. t.count .. ' ' .. t.classname)
+        end
+        print('with ' .. plural(empty_cages, 'empty cage'))
+    end
+end
+
+-- handle magic script arguments
+
+local args = {...}
+
+local list
+if args[2] == 'here' then
+    print "NOTE: The 'here' option isn't well tested for v50 and only works with the keyboard cursor."
+    local it = dfhack.gui.getSelectedItem(true)
+    list = {it}
+    if not df.item_cagest:is_instance(it) and
+        not df.item_animaltrapst:is_instance(it) then
+        list = {}
+        for _, cage in ipairs(df.global.world.items.other.ANY_CAGE_OR_TRAP) do
+            if same_xyz(df.global.cursor, cage.pos) then
```

Comment:
instead of the cursor, you could use the position of the selected item to find a cage at the same location

Reply:
Good idea since df.global.cursor currently provides garbage values unless the player uses the keyboard cursor.

Unfortunately, my naive fix that changed df.global.cursor to it.pos revealed a couple problems. Now that we're using 'it' (item) beyond checking if it's a valid cage or animal trap and then falling back on the keyboard cursor position if it's not, we need to make sure we select a valid item before using its position. Also we should definitely check to see if the player is selecting a placed cage, which counts as a building rather than an item, among other things.

I'll see if I can't fix this and make it more robust.

Reply:
> we should definitely check to see if the player is selecting a placed cage, which counts as a building rather than an item

Even if a cage has been placed, there should still also be an item present at the same position in the world.

> we need to make sure we select a valid item before using its position

Does `dfhack.gui.getSelectedItem()` not handle this? It should, but that may not be implemented yet.

Reply:
Ok, clarified on Discord that you mean the caller now needs to call `getSelectedItem()` and check if it actually returned something (previously, the cursor alone was used).